fix(twap): fix Fulfilled status detection

shoom3301 opened the pull request:
# Summary

Fixes #2985

The problem was here:
```
if (confirmedPartsCount === order.n) {
    if (+executionInfo.executedSellAmount > 0) return TwapOrderStatus.Fulfilled

    return TwapOrderStatus.Expired
  }
}
```

`confirmedPartsCount` - means the count of parts in the confirmed state (Expired, Cancelled, Fulfilled).
In the code above I wrongly assumed, that `confirmedPartsCount` === `all parts are filled`.
Because of it the status of an orders was wrongly detected as `Fulfilled`.

shoom3301 commented on src/modules/twap/utils/getTwapOrderStatus.test.ts:
```diff
@@ -0,0 +1,70 @@
+      const execution: TwapOrdersExecution = {
+        confirmedPartsCount: orderStruct.n,
+        info: {
+          executedSellAmount: (BigInt(orderStruct.partSellAmount) * BigInt(orderStruct.n)).toString(), // <----
```
I assume, since all parts are fill-or-kill, it's not possible, is it?

shoom3301 replied:
I also checked with fully filled TWAP order

alfetopito replied:
> I assume, since all parts are fill-or-kill, it's not possible, is it?

Technically yes.
But I've seen this happening.
Although I might be confusing with limit orders...

shoom3301 replied:
Then I have no idea how to fix it now :(

alfetopito replied:
There is already code in the project that deals with this:

https://github.com/cowprotocol/cowswap/blob/0ae6602f707045079c8a5e0f3b8d6bd70f703e61/src/utils/orderUtils/isOrderFilled.ts#L28

Be aware that your solution should work most of the time, I'm raising an edge case.

shoom3301 commented on src/modules/twap/utils/getTwapOrderStatus.ts:
```diff
@@ -11,19 +11,17 @@ export function getTwapOrderStatus(
 ): TwapOrderStatus {
   const isFulfilled = isTwapOrderFulfilled(order, executionInfo.executedSellAmount)
+  const isCancelled = auth === false
+  const isExpired = confirmedPartsCount === order.n || isTwapOrderExpired(order, executionDate)
```
For example, we have a TWAP order with 3 parts:
1. Expired
2. Cancelled
3. Fulfilled

The deadline for TWAP order is 13:00, the last part was filled at 12:30.
To not display a TWAP order as open for the last 30 minutes, we automatically mark it as Expired.

anxolin replied:
I also don't get it, if the order was fulfilled, it won't show as open, right, it will show as filled. Why in this case you would want to show expired?

shoom3301 replied:
<img width="1124" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/df867e2d-4c19-40f6-adb4-0a4271723cb8">

shoom3301 replied:
During this peach color timeline (12:25 - 12:30) there are no open parts and the parent is not fulfilled (only 1 part is filled). To not keep it open we mark it as expired.

alfetopito replied:
Thanks for the diagram, makes it much clearer!

As for the status.
It checks whether all of the orders have been confirmed (and that's the same as created on the backed?), but does not check their status. That's what's still not clear to me.

Or, does that mean: by now all orders that should have been created have already been created?

shoom3301 replied:
>and that's the same as created on the backed?

No, confirmed means status in [Expired, Cancelled, Fulfilled]. I've inherited the naming from the existing const `CONFIRMED_STATES`

shoom3301 replied:
So, the exact code:

```
export const CONFIRMED_STATES = [OrderStatus.FULFILLED, OrderStatus.EXPIRED, OrderStatus.CANCELLED, OrderStatus.FAILED]

const confirmedPartsCount = children.filter((order) => CONFIRMED_STATES.includes(order.status)).length
```

alfetopito replied:
Ah, got it!
Confirmed means (in my head) final states.
So when all parts are finalized, the `isExpired` is set to true, but it won't be a problem with fulfilled because it takes precedence, right?

elena-zh reviewed (changes requested):
Hey @shoom3301 , requesting changes due to: 
1. statuses are not auto-updated when a TWAP order should move from Signing to Open state; 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/c3622d25-194b-4695-9d74-cf0cdd19fff1)
when a part should be Expired: I need to refresh the browser in order to see these changes 
![hangs](https://github.com/cowprotocol/cowswap/assets/70885163/1a63280c-9fee-4093-b674-a89e4259e2e1)
2. When 2 parts were expired, and I run cancel TX for the past part, all parts are labeled as Cancelled (however, they are expired)
![cancelling](https://github.com/cowprotocol/cowswap/assets/70885163/9558048c-b46e-479f-8f86-179d06a31e38)

elena-zh commented:
One more issue: 
When the last part is filled, there are 3 states: 
1st MOO with Filled status, then parent bar is cleared out; the order goes to Open state
![Screenshot_1](https://github.com/cowprotocol/cowswap/assets/70885163/90c68e60-6b83-4f05-9043-8b82dba08983)
2nd moo: the order goes to the Filled state
3 step: all is displayed correctly again, but hangs in the Open orders tab until the order expires 
![3](https://github.com/cowprotocol/cowswap/assets/70885163/a7811a3d-2066-4693-b498-5e90870e6d51)

shoom3301 commented:
@elena-zh 

>statuses are not auto-updated when a TWAP order should move from Signing to Open state;

I believe it was a temporary issue with Safe

>when a part should be Expired: I need to refresh the browser in order to see these changes

Good point! Fixed. Now the time lag for updating status should be <= 5 sec

>When 2 parts were expired, and I run cancel TX for the past part, all parts are labeled as Cancelled (however, they are expired)
>In the end the 2 Expired parts are recorded as Cancelled, and the one Cancelled part is recorded as Expired

Fixed

>When the last part is filled, there are 3 states:
I can't reproduce it :(

elena-zh commented:
Hey @shoom3301 , case 2 from the https://github.com/cowprotocol/cowswap/pull/2987#pullrequestreview-1559145612 is not fixed :(

shoom3301 commented:
@elena-zh if it's the only issue left, then let's move it to another issue

elena-zh commented:
One more nitpick is that when the last part is filled, the order hangs in Open status. The parent order shows no filled status
![image](https://github.com/cowprotocol/cowswap/assets/70885163/daef0467-6000-4af5-b6ed-546faba15365)

The order hangs even if it's expired time is met, and until I refresh the page

shoom3301 commented:
@elena-zh fixed

elena-zh commented:
> One more nitpick is that when the last part is filled, the order hangs in Open status. The parent order shows no filled status ![image](https://user-images.githubusercontent.com/70885163/258120183-daef0467-6000-4af5-b6ed-546faba15365.png)
> 
> The order hangs even if it's expired time is met, and until I refresh the page

Hey @shoom3301 , not fixed :(

elena-zh commented:
Hey @shoom3301 , for cancelled statuses issue I have created #3007 bug. 
As for Filled/Expired orders not moving to the History tab, the issue is still here: 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/d25cda28-0461-4b2c-af41-ba784bdfb7d1)

shoom3301 commented:
@elena-zh I've found the issue :)
It's only reproducible in wrap/unwrap mode. In this case, twap orders data doesn't get updated.
Fixed that.

alfetopito commented on src/modules/twap/hooks/useTwapOrdersExecutions.ts:
```diff
@@ -68,11 +71,63 @@ export function useTwapOrdersExecutions(ids: string[]): TwapOrdersExecutionMap {
+ * In this case the discreteOrders array will contain only one order (the last one)
+ * So, we need to calculate the number of confirmed parts based on the time passed
+ */
+function getConfirmedPartsCount(twapOrderInfo: TwapOrderInfo, discreteOrders: Order[]): number {
```
This is all a bit complex, but I can understand it.
The comments are very helpful

shoom3301 replied:
The case is a bit complex :)

alfetopito commented on src/modules/twap/updaters/TwapOrdersUpdater.tsx:
```diff
@@ -79,6 +78,7 @@ export function TwapOrdersUpdater(props: {
       twapOrderExecutions.current
     )
 
+    console.debug('[TwapOrdersUpdater] Update TWAP orders', items)
```
Are we keeping this?

shoom3301 replied:
Thanks! Removed

alfetopito commented on src/pages/AdvancedOrders/index.tsx:
```diff
@@ -27,11 +27,10 @@ export default function AdvancedOrdersPage() {
       <styledEl.PageWrapper isUnlocked={isUnlocked}>
         <styledEl.PrimaryWrapper>
-          <AdvancedOrdersWidget>
+          <AdvancedOrdersWidget updaters={<TwapUpdaters />}>
```
Question: would it work if you just pass the component import rather than "instantiating" it here, like this:

```suggestion
          <AdvancedOrdersWidget updaters={TwapUpdaters}>
```

I haven't tested, just curiosity

shoom3301 replied:
<img width="960" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/9e9a57b9-1edc-4c24-bd44-1b540bf2d80b">

Nope

elena-zh reviewed (approved):
Hey @shoom3301 , 'expired' and 'fileld' cases LGTM!
As for the 'cancel' last part case, I have created #3007 issue for this.
